### ISSUE TITLE
Internal methods are still weak

### ISSUE THREAD

**Issue by resistorsoftware** [Shopify/shopify-api-ruby#566]
Going back 10 years now... and still no progress on this. 

A custom collection has many products. Let us say 119 of them. So we use the API to get at that custom collection using a GET. There is a method **products** that is supposed to return all the products in the collection, but it is and has always been, pinned to just 50. 

So if we ever write code trying to pretend we are working with all the products this will fail. Instead we have to override the products method and include one that pages through them all and therefore acts as one would think this method would act.

Could Shopify maybe finally address this **weakness**?

**Comment by katiedavis** [comment]
Hi @resistorsoftware, thanks for filing an issue. I believe the issue here is that the CustomCollections result is paginated and this isn't well documented.

I'm filling an issue internally to update our help documents to provide more clarity around this.

The new GraphQL APIs also provide cleaner pagination support, if you're looking for a solution for handling pagination.

**Comment by resistorsoftware** [comment]
Actually, paging with GraphQL is not great. It is really easy to get messed up, especially if you a trying to use the GraphQL cursors to move forward and backward. Have you ever actually tried that and had success? I know when I was playing with it, it was not working as expected. But I digress. Also, I see nothing in the shopify_api code that supports paging under the hood. Can you point me to the relevant lines of code in the gem for that please. I am interested to see what I have obviously missed.

**Comment by katiedavis** [comment]
Hi again,
`shopify_api` does not handle pagination, so you would have to handle that on your end. I asked around for best practices for pagination with our REST APIs and this is what I found:
```
the _recommended_ approach is to use the `since_id` parameter and *not* `page`. `since_id` requires the partner to keep track of the maximum resource id returned in each call to then set as the `since_id` parameter in the subsequent call. It’s recommended as it’s significantly more performant than `page`, which can actually lead to database timeouts when dealing with large datasets.
```
I'd like to thank you for bringing this issue up. We're exploring options for adding pagination into the shopify_api in the future as well.

**Comment by resistorsoftware** [comment]
I am not sure how to demonstrate my issue but I suppose I will give it one more shot, and for posterity, anyone reading this will understand where I am coming from. 

The shopify_api endpoint where you GET a resource called CustomCollection returns that **Custom Collection**. An interesting attribute Shopify provide is the **products_count**. When you GET a collection, you can tell how many products are in it. So let us assume there are > 50.

Fast forward to using the gem in a REPL and using the method someone dropped in long ago. That method is **products**.

So I can do the following one-two punch:

1. my_collection = ShopifyAPI::CustomCollection.first
2. my_collection.products.each { |product| puts product.title }

And I will get only 50 products even if the collection has 100 products. And note, there is no way to paginate that method, or otherwise get at all 100 products unless you override the products method with your own paging code. Which I do... but which is not really convenient. And this problem is littered all over the RestAPI and has been for some time. And since Shopify is not likely to abandon the RestAPI for now, I am just putting it out there, that these small niggling things, continue to niggle. 

By the way, this has nothing to do with a **since_id** either. That would be appropriate when asking for orders, but not the products in a custom collection.

**Comment by nwtn** [comment]
Thanks for the detailed description and feedback @resistorsoftware. We’ll set aside some time to look at this shortly.

Would you be able to describe how you’re doing the pagination in your overrides?

**Comment by resistorsoftware** [comment]
Any method can be overridden in Ruby. So I just do that by re-opening the API gem class.

    module ShopifyAPI
        class CustomCollection
           def products
              puts "Paging through the Products belonging to the Custom Collection: #{self.id}"
              # do paging so we get all the product. Note that this is crazy as it could take some time
              # return more than 50 products when there are more than 50, unlike the shopify method
          end
       end
    end

**Comment by nwtn** [comment]
Oh I know any method can be overridden :D
I was curious to see your actual paging code, and whether there was anything special/unusual you had to do to get it working.

**Comment by resistorsoftware** [comment]
Nothing special... way back in the day, in when caveman roamed the earth, we discussed how it would be neat if the API methods like this responded to paging and limit params like their parents. But alas, no meat was found in the neighborhood, the local yokels took off for greener climes, and we are nearly a decade later with the same issues for underlying resources. Stuck at 50 for the unwary. 

Clearly this is a tough issue, and don't tell me it all goes away with GraphQL as I know that is also not true at all. Just a fact that this comes up, and while it may be few and far between for vanilla merchants selling t-shirts, I have some clients doing sophisticated things with collections, and collections are very weakly supported. Heck they don't even sort themselves on Shopify! as I have had to fix that too.

Any... sample paging code that works but is far from optimal, as it has no parameters to help it along:

      products = []
      page = 1
      count = Product.count(collection_id: self.id)
      return unless count > 0
      page += count.divmod(250).first
      status = {
        limit: 250,
        collection_id: self.id
      }
      while page > 0
        status[:page] = page
        products += Product.all(params: status)
        page -= 1
      end
      products
    end
